client/v3/config.go

```diff
@@ -90,3 +93,103 @@ type Config struct {
 
 	// TODO: support custom balancer picker
 }
+
+type ClientConfig struct {
+	Endpoints        []string      `json:"endpoints"`
+	RequestTimeOut   time.Duration `json:"request-timeout"`
+	DialTimeout      time.Duration `json:"dial-timeout"`
+	KeepAliveTime    time.Duration `json:"keepalive-time"`
+	KeepAliveTimeout time.Duration `json:"keepalive-timeout"`
+	Scfg             *SecureCfg    `json:"secure-cfg"`
+	Acfg             *AuthCfg      `json:"auth-cfg"`
+}
+
+type SecureCfg struct {
+	Cert       string `json:"cert"`
+	Key        string `json:"key"`
+	Cacert     string `json:"cacert"`
+	ServerName string `json:"servername"`
+
+	InsecureTransport  bool `json:"insecure-transport"`
+	InsecureSkipVerify bool `json:"insecure-skip-tls-verify"`
+}
+
+type AuthCfg struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+func (cc *ClientConfig) MustClient() *Client {
+	cfg, err := NewClientCfg(cc.Endpoints, cc.DialTimeout, cc.KeepAliveTime, cc.KeepAliveTimeout, cc.Scfg, cc.Acfg)
+	if err != nil {
+		cobrautl.ExitWithError(cobrautl.ExitBadArgs, err)
+	}
+
+	client, err := New(*cfg)
+	if err != nil {
+		cobrautl.ExitWithError(cobrautl.ExitBadConnection, err)
+	}
+
+	return client
+}
+
+func NewClientCfg(endpoints []string, dialTimeout, keepAliveTime, keepAliveTimeout time.Duration, scfg *SecureCfg, acfg *AuthCfg) (*Config, error) {
```

Comment:
Would recommend constructor only be responsible for providing default values and not setting or validating. For configuration structs I expect values to be provided via `json.Unmarshall` or `flags.VarString`. Both of those functions take reference to struct/field making the constructor you wrote not very useful.

Example on how to define configuration (based on how K8s code does configuration):
```
type ClientConfig struct {
  DialTimeout time.Duration `json:"dial-timeout"`
}

func NewClientConfig() ClientConfig {
  return ClientConfig{
    DialTimeout: 5*time.Second,
  }
}

func (c ClientConfig) Validate() err {
  if c.DialTimeout < 0 {
    return fmt.Error("Timeout cannot be negative")
  }
  return nil
}

func (c ClientConfig) AddFlags(fs flags.FlagSet) {
  fs.VarString("dial-timeout", &c.DialTimeout, c.DialTimeout, "Dial timeout")
}

func main() {
  
  config := NewClientConfig()
  config.AddFlags(flags.CommandLine) // or something like
  err := config.Validate()
  if err != nil {
    fmt.Print(err)
    os.Exit(1)
  }
}
```

This code is very draftly, however shows overall direction. As for registering single `ClientConfig` as different flags we could use aliasing. For example:

```
func addClientConfigFlags(c ClientConfig, prefix string, fs flags.FlagSet) {
  fs.VarString(prefix + "-dial-timeout", &c.DialTimeout, c.DialTimeout, "Dial timeout")
}

type DiscoveryClientConfig ClientConfig
type EtcdctlClientConfig ClientConfig

func (c DialClientConfig) AddFlags(fs flags.FlagSet) {
  addClientConfigFlags(c, "discovery", fs)
}

func (c EtcdctlClientConfig) AddFlags(fs flags.FlagSet) {
  addClientConfigFlags(c, "client", fs)
}
```